ops/log.py

```diff
@@ -87,5 +91,37 @@ def except_hook(etype: type[BaseException], value: BaseException, tb: types.Trac
         if exc_stderr:
             print(f'Uncaught {etype.__name__} in charm code: {value}', file=sys.stderr)
             print('Use `juju debug-log` to see the full traceback.', file=sys.stderr)
+        _security_event(
+            f'sys_crash:{etype.__name__}',
+            level='ERROR',
+            description=f'Uncaught exception in charm code: {value!r}.',
+        )
 
     sys.excepthook = except_hook
+
+
+def _security_event(event: str, *, level: str, description: str):
+    """Send a structured security event log to Juju, as defined by SEC0045.
+
+    Args:
+        event: the name of the event, in the format described by OWASP
+          https://cheatsheetseries.owasp.org/cheatsheets/Logging_Vocabulary_Cheat_Sheet.html
+        level: log level, such as 'DEBUG', 'INFO', or 'ERROR'
+        description: a free-form description of the event, meant for human
+          consumption. Includes additional details of the event that do not
+          fit in the event name.
+    """
+    logger = logging.getLogger(__name__)
+    data: dict[str, typing.Any] = {
+        # This duplicates the timestamp that will be in the Juju log, but is
+        # included so that applications that are pulling out only the structured
+        # data can still see the time of the event.
+        'datetime': datetime.datetime.now(datetime.timezone.utc).isoformat(),
+        # Similarly, this duplicates the level at which this is logged to Juju.
+        'level': level,
+        'type': 'security',
+        'appid': os.environ.get('JUJU_MODEL_UUID', 'unknown'),
```

Comment:
I think you pointed this out when we talked, but this is called "appid" but is actually the model UUID. Should we use something else?

Also: it's unfortunate we have to use the environment variable here, but I guess there's not really a cleaner way.

Could we have `app_id` be an optional parameter, and if not set, it uses the environment variable? (See comment below about Pebble logging too.)